### PR TITLE
Implemented Lock/Unlock Functionality for Issues

### PR DESCRIPTION
**Pull request by prayankmathur**
@shiftkey 
I have added the unit and integration tests for issue #896 . however the build is not running successfully.

### CODE REVIEW

**Comment by prayankmathur** [comment]
@shiftkey 

The build passed the checks here i guess.
Take a look at the unit and integration tests.

**Comment by ryangribble** [comment]
Hi @prayankmathur , could I suggest that you reword the pull request title and description to contain a little bit more details about what it's about?  
You certainly want to link to the issue #896 but also would be good to say a few words here about it so everyone doesnt have to click through to the issue to see what this PR is about.

eg something like this:

Title:
> Implement Lock/Unlock functionality for Issues

Description:
> This PR implements the locking and unlocking funcitonality for Issues
Fixes #896
Also includes Unit and Integration Tests


Also from a quick look, this pull request includes your changes for ther other PR about ItemState/ItemStateFilter etc, which makes it hard to review the diffs in isolation.  Ideally you would either do the work on separate branches, or once one PR gets merged, you can merge master into your branch and we will then see the "true" differences in this PR

**Comment by prayankmathur** [comment]
@ryangribble

Changed the title. Sorry for the trouble. Actually i did branch out this issue, but it turned out that i branched it from my master branch which consisted of the previous bug(which i hadn't branched).
If you could take the trouble of going through the other PR first, then it could solve the issue.

BTW, is there any problem with this [function](https://github.com/octokit/octokit.net/blob/master/Octokit.Tests/Reactive/ObservableIssuesClientTests.cs#L327)

**Comment by prayankmathur** [comment]
@ryangribble 
@shiftkey 
I found that this [function](http://github.com/octokit/octokit.net/blob/master/Octokit.Tests/Reactive/ObservableIssuesClientTests.cs#L327) was not correct.
So corrected that too in the same PR.
Hope that it is fine.

**Comment by shiftkey** [comment]
@prayankmathur given it's not related to this feature, a separate PR would be nice for a couple of reasons:

 - we can review it separate to this
 - given it's a simpler change, it's likely to get merged in quicker

I'll leave it up to you whether you'd like to extract that to a new branch.

**Comment by prayankmathur** [comment]
@shiftkey 
I ll open a separate PR.

**Comment by ryangribble** [comment]
Just noting for reference that the status of this PR is
- Waiting for #1140 to be merged and then have this PR's commits tidied up (due to this PR being built ontop of #1140)
- Waiting on @prayankmathur to split out the test fix for `UpdateIssue` to another PR and remove the commit from this one

**Comment by prayankmathur** [comment]
@ryangribble 
I was waiting for #1140 to be merged.
But now i have opened a separate PR for the UpdateMethod function correction.

**Comment by shiftkey** [comment]
@prayankmathur this looks like it contains work from #1140 (open) and #1196 (merged). Are you able to rebase the branch so that this only contains the work related to the locking and unlocking of issues?

File: Octokit.Tests/Reactive/ObservableIssuesClientTests.cs [hunk 1]
```diff
@@ -329,11 +329,61 @@ public void EnsuresArgumentsNotNull()
...
-            Assert.Throws<ArgumentException>(() => client.Create("", "name", new NewIssue("x")));
-            Assert.Throws<ArgumentNullException>(() => client.Create("owner", null, new NewIssue("x")));
-            Assert.Throws<ArgumentException>(() => client.Create("owner", "", new NewIssue("x")));
-            Assert.Throws<ArgumentNullException>(() => client.Create("owner", "name", null));
+            Assert.Throws<ArgumentNullException>(() => client.Update(null, "name", 42, new IssueUpdate()));
+            Assert.Throws<ArgumentException>(() => client.Update("", "name", 42, new IssueUpdate()));
+            Assert.Throws<ArgumentNullException>(() => client.Update("owner", null, 42, new IssueUpdate()));
+            Assert.Throws<ArgumentException>(() => client.Update("owner", "", 42, new IssueUpdate()));
+            Assert.Throws<ArgumentNullException>(() => client.Update("owner", "name", 42, null));
+        }
```
Comment (prayankmathur) [inline Octokit.Tests/Reactive/ObservableIssuesClientTests.cs]: @shiftkey 
Why is this function showing changes.
I had opened a seperate PR to change it which has been merged.
The function has changed also [here](https://github.com/octokit/octokit.net/blob/master/Octokit.Tests/Reactive/ObservableIssuesClientTests.cs#L327)

**Comment by prayankmathur** [comment]
@shiftkey 
I have made the necessary changes. Please take a look.

File: Octokit.Tests.Integration/Clients/IssuesClientTests.cs [hunk 2]
```diff
@@ -62,6 +62,26 @@ public IssuesClientTests()
     }
 
     [IntegrationTest]
+    public async Task CanLockAndUnlockIssue()
+    {
+         var newIssue = new NewIssue("a test issue") { Body = "A new unassigned issue" };
+         var issue = await _issuesClient.Create(_context.RepositoryOwner, _context.RepositoryName, newIssue);
+         var retrieved = await _issuesClient.Get(_context.RepositoryOwner, _context.RepositoryName, issue.Number);
+ 
+         Assert.Equal(false, issue.Locked);
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Clients/IssuesClientTests.cs]: You can use `Assert.False` or `Assert.True` here to make these a bit more succinct.

File: Octokit.Tests.Integration/Reactive/ObservableIssuesClientTests.cs [hunk 3]
```diff
@@ -72,6 +72,23 @@ public ObservableIssuesClientTests()
         Assert.Equal("Modified integration test issue", updateResult.Title);
     }
 
+     [IntegrationTest]
+     public async Task CanLockAndUnlockIssues()
+     {
+         var newIssue = new NewIssue("Integration Test Issue");
+ 
+         var createResult = await _client.Create(_context.RepositoryOwner, _context.RepositoryName, newIssue);
+         Assert.Equal(false, createResult.Locked);
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Reactive/ObservableIssuesClientTests.cs]: `Assert.True` or `Assert.False` in this test too

File: Octokit/Clients/IIssuesClient.cs [hunk 4]
```diff
@@ -163,5 +163,25 @@ public interface IIssuesClient
...
+
+        /// <summary>
+        /// Locks an issue for the specified repository. Issue owners and users with push access can lock an issue.
+        /// </summary>
+        /// <remarks>https://developer.github.com/v3/issues/#lock-an-issue</remarks>
+        /// <param name="owner">The owner of the repository</param>
+        /// <param name="name">The name of the repository</param>
+        /// <param name="number">The issue number</param>
+        /// <returns></returns>
+        Task LockIssue(string owner, string name, int number);
```
Comment (shiftkey) [inline Octokit/Clients/IIssuesClient.cs]: Given this usage will look like this:

```
await github.Issue.LockIssue("owner", "name", 1);
```

Perhaps we can drop the `Issue` suffix here, to give:

```
github.Issue.Lock("owner", "name", 1);
```

Thoughts @haacked @ryangribble?

File: Octokit/Clients/IssuesClient.cs [hunk 5]
```diff
@@ -225,5 +225,37 @@ public Task<Issue> Update(string owner, string name, int number, IssueUpdate iss
...
+        /// <param name="owner">The owner of the repository</param>
+        /// <param name="name">The name of the repository</param>
+        /// <param name="number">The issue number</param>
+        /// <returns></returns>
+        public Task LockIssue(string owner, string name, int number)
+        {
+            Ensure.ArgumentNotNullOrEmptyString(owner, "owner");
+            Ensure.ArgumentNotNullOrEmptyString(name, "name");
+
+            return ApiConnection.Put(ApiUrls.IssueLock(owner, name, number));
```
Comment (shiftkey) [inline Octokit/Clients/IssuesClient.cs]: As per the [blog post](https://developer.github.com/changes/2016-02-11-issue-locking-api/) we need to pass in a custom `Accept` header here (as well as for the `Unlock` action below)

File: Octokit/Clients/IIssuesClient.cs (hunk 4 above)
Comment (ryangribble) [inline Octokit/Clients/IIssuesClient.cs]: Yep I like the sound of `.Lock()` and `.Unlock()` :+1:

File: Octokit/Clients/IssuesClient.cs (hunk 5 above)
Comment (ryangribble) [inline Octokit/Clients/IssuesClient.cs]: Yep, see [AcceptHeaders.cs`(https://github.com/octokit/octokit.net/blob/master/Octokit/Helpers/AcceptHeaders.cs) helper class and other usages, for how we can do this

**Comment by prayankmathur** [comment]
@shiftkey @ryangribble 
Should i rename the reactive function also ?
Because then it's usage would be like
>client.Lock(owner, name, 1);

Currently its like 
>client.LockIssue(owner, name, 1)

**Comment by ryangribble** [comment]
Yes you should rename it too.  In integration or unit tests, you are directly referencing the `IObserableIssuesClient` but just remember that "regular" users of octokit will be accessing both the standard or observable client, through the "proper" heierarchy
eg `ObservableGitHubClient.Issue.Lock()`

**Comment by prayankmathur** [comment]
@ryangribble 
Yeah, forgot that :grin: 
will make the changes.

**Comment by ryangribble** [comment]
Also dont forget you've got merge conflicts so either merge master into your branch, or rebase your branch on master and force push

**Comment by prayankmathur** [comment]
@ryangribble 
I resolved the issues too and have made the changes. Please take a look.

File: Octokit/Helpers/AcceptHeaders.cs [hunk 6]
```diff
@@ -14,6 +14,9 @@ public static class AcceptHeaders
 
         public const string StarCreationTimestamps = "application/vnd.github.v3.star+json";
 
+        public const string IssueLockingUnlockingApiPreview = "application/vnd.github.the-key-preview+json";
+
         public const string CommitReferenceSha1Preview = "application/vnd.github.chitauri-preview+sha";
+
```
Comment (ryangribble) [inline Octokit/Helpers/AcceptHeaders.cs]: Extra whitespace line introduced here :)

File: Octokit.Tests/Reactive/ObservableIssuesClientTests.cs [hunk 7]
```diff
@@ -337,6 +337,56 @@ public void EnsuresArgumentsNotNull()
         }
     }
 
+    public class TheLockIssueMethod
```
Comment (ryangribble) [inline Octokit.Tests/Reactive/ObservableIssuesClientTests.cs]: I guess this now should be "TheLockMethod"

File: Octokit.Tests/Reactive/ObservableIssuesClientTests.cs [hunk 8]
```diff
@@ -337,6 +337,56 @@ public void EnsuresArgumentsNotNull()
...
+            var client = new ObservableIssuesClient(gitHubClient);
+
+            Assert.Throws<ArgumentNullException>(() => client.LockIssue(null, "name", 42));
+            Assert.Throws<ArgumentException>(() => client.LockIssue("", "name", 42));
+            Assert.Throws<ArgumentNullException>(() => client.LockIssue("owner", null, 42));
+            Assert.Throws<ArgumentException>(() => client.LockIssue("owner", "", 42));
+        }
+    }
+
+    public class TheUnlockIssueMethod
```
Comment (ryangribble) [inline Octokit.Tests/Reactive/ObservableIssuesClientTests.cs]: "TheUnlockMethod"

File: Octokit.Tests/Reactive/ObservableIssuesClientTests.cs [hunk 9]
```diff
@@ -337,6 +337,56 @@ public void EnsuresArgumentsNotNull()
...
+    {
+        [Fact]
+        public void UnlockIssue()
```
Comment (ryangribble) [inline Octokit.Tests/Reactive/ObservableIssuesClientTests.cs]: Wording for the unit tests should read TheBlahMethod....DoesSomething

So...
TheUnlockMethod->Unlock**s**Issue

(note Unlocks rather than Unlock)

File: Octokit.Tests/Reactive/ObservableIssuesClientTests.cs [hunk 10]
```diff
@@ -337,6 +337,56 @@ public void EnsuresArgumentsNotNull()
         }
     }
 
+    public class TheLockIssueMethod
+    {
+        [Fact]
+        public void LockIssue()
```
Comment (ryangribble) [inline Octokit.Tests/Reactive/ObservableIssuesClientTests.cs]: "Lock**s**Issue"

**Comment by ryangribble** [comment]
Looks like you're ontop of the test failures which is :+1: 

Ive highlighted a couple of REALLY minor consistency issues around naming unit tests (now the methods have been renamed)

**Comment by prayankmathur** [comment]
@ryangribble 
The changes are made successfully. You can now finally take a look. :grinning:

File: Octokit.Tests.Integration/Clients/IssuesClientTests.cs [hunk 11]
```diff
@@ -61,6 +61,25 @@ public IssuesClientTests()
...
+    {
+        var newIssue = new NewIssue("a test issue") { Body = "A new unassigned issue" };
+        var issue = await _issuesClient.Create(_context.RepositoryOwner, _context.RepositoryName, newIssue);
+        var retrieved = await _issuesClient.Get(_context.RepositoryOwner, _context.RepositoryName, issue.Number);
+        Assert.False(issue.Locked);
+ 
+        await _issuesClient.Lock(_context.RepositoryOwner, _context.RepositoryName, issue.Number);
+        retrieved = await _issuesClient.Get(_context.RepositoryOwner, _context.RepositoryName, issue.Number);
+        Assert.NotNull(retrieved);
+        Assert.True(issue.Locked);
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Clients/IssuesClientTests.cs]: We should be checking `retrieved.Locked` here, right?

File: Octokit.Tests.Integration/Clients/IssuesClientTests.cs [hunk 12]
```diff
@@ -61,6 +61,25 @@ public IssuesClientTests()
...
+ 
+        await _issuesClient.Unlock(_context.RepositoryOwner, _context.RepositoryName, issue.Number);
+        retrieved = await _issuesClient.Get(_context.RepositoryOwner, _context.RepositoryName, issue.Number);
+        Assert.NotNull(retrieved);
+        Assert.False(issue.Locked);
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Clients/IssuesClientTests.cs]: `retrieved.Locked` here too

File: Octokit.Tests.Integration/Clients/IssuesClientTests.cs [hunk 13]
```diff
@@ -61,6 +61,25 @@ public IssuesClientTests()
...
     [IntegrationTest]
+    public async Task CanLockAndUnlockIssue()
+    {
+        var newIssue = new NewIssue("a test issue") { Body = "A new unassigned issue" };
+        var issue = await _issuesClient.Create(_context.RepositoryOwner, _context.RepositoryName, newIssue);
+        var retrieved = await _issuesClient.Get(_context.RepositoryOwner, _context.RepositoryName, issue.Number);
+        Assert.False(issue.Locked);
+ 
+        await _issuesClient.Lock(_context.RepositoryOwner, _context.RepositoryName, issue.Number);
+        retrieved = await _issuesClient.Get(_context.RepositoryOwner, _context.RepositoryName, issue.Number);
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Clients/IssuesClientTests.cs]: Actually, you should be returned the updated `Issue` from `Lock` and `Unlock` here - to simplify this test (and do less network calls) you could assign this returned value to `retrieved`...

File: Octokit/Clients/IssuesClient.cs [hunk 14]
```diff
@@ -225,5 +225,37 @@ public Task<Issue> Update(string owner, string name, int number, IssueUpdate iss
...
+        /// <param name="owner">The owner of the repository</param>
+        /// <param name="name">The name of the repository</param>
+        /// <param name="number">The issue number</param>
+        /// <returns></returns>
+        public Task<Issue> Lock(string owner, string name, int number)
+        {
+            Ensure.ArgumentNotNullOrEmptyString(owner, "owner");
+            Ensure.ArgumentNotNullOrEmptyString(name, "name");
+
+            return ApiConnection.Put<Issue>(ApiUrls.IssueLock(owner, name, number), AcceptHeaders.IssueLockingUnlockingApiPreview);
```
Comment (shiftkey) [inline Octokit/Clients/IssuesClient.cs]: This is not quite the right overload, and sadly `ApiConnection` doesn't quite have an easy way to do this.

This works for me:

```
return ApiConnection.Put<Issue>(ApiUrls.IssueLock(owner, name, number), new object(), null, AcceptHeaders.IssueLockingUnlockingApiPreview);
```

File: Octokit/Clients/IIssuesClient.cs [hunk 15]
```diff
@@ -163,5 +163,25 @@ public interface IIssuesClient
...
+
+        /// <summary>
+        /// Locks an issue for the specified repository. Issue owners and users with push access can lock an issue.
+        /// </summary>
+        /// <remarks>https://developer.github.com/v3/issues/#lock-an-issue</remarks>
+        /// <param name="owner">The owner of the repository</param>
+        /// <param name="name">The name of the repository</param>
+        /// <param name="number">The issue number</param>
+        /// <returns></returns>
+        Task<Issue> Lock(string owner, string name, int number);
```
Comment (shiftkey) [inline Octokit/Clients/IIssuesClient.cs]: This endpoint doesn't return any data, so it needs to return a `Task`.

File: Octokit/Clients/IssuesClient.cs [hunk 16]
```diff
@@ -225,5 +225,37 @@ public Task<Issue> Update(string owner, string name, int number, IssueUpdate iss
...
+        /// <param name="owner">The owner of the repository</param>
+        /// <param name="name">The name of the repository</param>
+        /// <param name="number">The issue number</param>
+        /// <returns></returns>
+        public Task Unlock(string owner, string name, int number)
+        {
+            Ensure.ArgumentNotNullOrEmptyString(owner, "owner");
+            Ensure.ArgumentNotNullOrEmptyString(name, "name");
+
+            return ApiConnection.Delete(ApiUrls.IssueLock(owner, name, number), AcceptHeaders.IssueLockingUnlockingApiPreview);
```
Comment (ryangribble) [inline Octokit/Clients/IssuesClient.cs]: > he name is the least-worst option that came to mind given it'll clash with the other

The inconsistency of overloads for Post Put Delete and Get makes me :cry: too!

Rather than introduce the first/only occurence of this "{verb}With" terminology to get around the overload collisions, perhaps in this case we could just go with adding an overload for
`Delete(Uri uri, object data, string accepts)`
and call it as `ApiConnection.Delete(ApiUrls.IssueLock(owner, name, number), null, AcceptHeaders.IssueLockingUnlockingApiPreview)`
To me this seems slightly less ugly than adding in a new terminology

Or alternatively, perhaps we could introduce a whole new set of "With" overloads, and start to obsolete the old ones

eg

```
GetWithCustomAccept()
PutWithCustomAccept()
DeleteWithCustomAccept()
PostWithCustomAccept()
```

File: Octokit/Clients/IssuesClient.cs (hunk 14 above)
Comment (ryangribble) [inline Octokit/Clients/IssuesClient.cs]: Does the body have to be `new object()` or can it just be `null` ?

File: Octokit/Clients/IIssuesClient.cs (hunk 15 above)
Comment (ryangribble) [inline Octokit/Clients/IIssuesClient.cs]: Given that the API does return a HTTP 204 NoContent response when "succesful" the way ive seen this implemented in octokit in other places (eg [OrganizationMembers.Publicize()](https://github.com/octokit/octokit.net/blob/829677c9292ca503806aac385f49e37ad3730fdd/Octokit/Clients/OrganizationMembersClient.cs#L289)) is that the Octokit call returns a Task<bool> then internally it calls the `Connection.Put` and checks the HTTP response.  If a 204 was received then "true" is returned, otherwise false (or an exception of course, for 404/500/other stuff handled by the plumbing inside Connection class)

**Comment by ryangribble** [comment]
Regarding the actual implementation of Lock and Unlock methods (which are API PUT methods that return a 204 response on success) I suggest we should implement them similarly to  [OrganizationMembers.Publicize()](https://github.com/octokit/octokit.net/blob/829677c9292ca503806aac385f49e37ad3730fdd/Octokit/Clients/OrganizationMembersClient.cs#L289)

In other words, check for that 204 response and return a `Task<bool>` to indicate success

**Comment by prayankmathur** [comment]
@ryangribble 
@shiftkey 
Right now, I am implementing the >Delete(Uri uri, object data, string accepts)
And then later on we can open up a new PR.
Is that Fine ?

File: Octokit/Clients/IIssuesClient.cs (hunk 15 above)
Comment (shiftkey) [inline Octokit/Clients/IIssuesClient.cs]: :thumbsup:

File: Octokit/Clients/IssuesClient.cs (hunk 16 above)
Comment (shiftkey) [inline Octokit/Clients/IssuesClient.cs]: > we could just go with adding an overload for `Delete(Uri uri, object data, string accepts)` and then pass null as the body

I don't think we can pass `null` here, but everything else is :thumbsup: from me

**Comment by prayankmathur** [comment]
@shiftkey 
@ryangribble 
Just a gentle reminder to go see the changes that i made in this PR.

**Comment by ryangribble** [comment]
> Also should I implement the Lock/Unlock methods as suggested by @ryangribble which would return 
Task<bool>

Yes please :grinning: @shiftkey also gave this a :+1: so we are in agreement that when the API docs mention a 204 response as success... you can check for that and return `Task<bool>` accordingly.  Link to existing example: [OrganizationMembers.Publicize()](https://github.com/octokit/octokit.net/blob/829677c9292ca503806aac385f49e37ad3730fdd/Octokit/Clients/OrganizationMembersClient.cs#L289)

File: Octokit/Http/Connection.cs [hunk 17]
```diff
@@ -485,6 +485,22 @@ Task<IApiResponse<T>> SendDataInternal<T>(object body, string accepts, string co
         }
 
         /// <summary>
+        /// Performs an asynchronous HTTP DELETE request that expects an empty response.
+        /// </summary>
+        /// <param name="uri">URI endpoint to send request to</param>
+        /// <param name="data">The object to serialize as the body of the request</param>
+        /// <param name="accepts">Specifies accept response media type</param>
+        /// <returns>The returned <seealso cref="HttpStatusCode"/></returns>
+        public async Task<HttpStatusCode> Delete(Uri uri,object data, string accepts)
```
Comment (ryangribble) [inline Octokit/Http/Connection.cs]: missing space in between uri and object

File: SolutionInfo.cs [hunk 18]
```diff
@@ -6,10 +6,8 @@
 [assembly: AssemblyVersionAttribute("0.19.0")]
 [assembly: AssemblyFileVersionAttribute("0.19.0")]
 [assembly: ComVisibleAttribute(false)]
-namespace System
-{
-    internal static class AssemblyVersionInformation
-    {
+namespace System {
+    internal static class AssemblyVersionInformation {
```
Comment (ryangribble) [inline SolutionInfo.cs]: Is this change intended?  Probably best to revert this one

File: Octokit/Http/Connection.cs (hunk 17 above)
Comment (shiftkey) [inline Octokit/Http/Connection.cs]: This will be handled by the next `FormatCode` call, so let's :boot: addressing this...

File: SolutionInfo.cs (hunk 18 above)
Comment (shiftkey) [inline SolutionInfo.cs]: This is the eternal struggle between the FAKE script and `FormatCode` - see https://github.com/shiftkey/Octokit.CodeFormatter/issues/2 for how I'd like to tackle this...

**Comment by shiftkey** [comment]
I'm happy with the integration test now passing, and the minor stuff remaining can be sorted out later.

Thanks @prayankmathur!

**Comment by prayankmathur** [comment]
@shiftkey
:+1:
